num/quat/trig.go

```diff
@@ -0,0 +1,143 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
```

Comment:
Do you have a reference for these functions?

Reply:
These were in the original ["math/quat" CL](https://go-review.googlesource.com/c/39220/1). Other than that, no I don't (and yes it does trouble me that the only reference is an April fools joke).

Reply:
@mdempsky Do you remember where you got the references for the quaternion [April fools joke for 2017](https://go-review.googlesource.com/c/39220/1)? In particular, the trig functions.

Reply:
I'd be skeptical of code from that CL. My intention was primarily to inspire a "wait, is this serious?" reaction, and not expecting any of the code would ever be used in production.

I don't remember the exact origin of the code either. Trying to retrace my steps, what I do vaguely remember at least is reading Dorai Sitaram's "On adding quaternions to Scheme" (http://www.ccs.neu.edu/home/dorai/squat/squat.html), and comparing it with Go's math/cmplx package. I believe the math/quat package was largely my own best-guess extrapolation from those two sources.

Reply:
Thanks. That scheme reference does look promising.

Interestingly, while the remainder of the math/quat looks good as far as I can tell, there was an error in the quaternion multiplcation code.

I like the idea that an April fools joke will get into a codebase — properly vetted.

Reply:
> there was an error in the quaternion multiplcation code.

Out of curiosity, what/where was this error?

Reply:
The error was in the mul node [here](https://go-review.googlesource.com/c/go/+/39216/1/src/go/constant/value.go#1266). From memory, it gives the wrong sign sometimes.

Reply:
@btracey I've had a look through the Scheme paper and what it covers (circular trig functions) looks OK.

For inverse trig functions, there is [this chapter](https://link.springer.com/content/pdf/10.1007%2F978-3-0348-0622-0_8.pdf) from [Real Quaternionic Calculus Handbook](https://www.springerprofessional.de/en/real-quaternionic-calculus-handbook/1981012) which I don't have access to via our library for other chapters, although Google books has a [reasonable coverage of it](https://books.google.com/books?id=YnS8BAAAQBAJ&printsec=frontcover&source=gbs_ge_summary_r&cad=0#v=onepage&q&f=false) (but not hyperbolic trig).

Reply:
@btracey I've had a look through the Scheme paper and what it covers
(circular trig functions) looks OK.

For inverse trig functions, there is [this chapter](https://link.springer.com/
content/pdf/10.1007%2F978-3-0348-0622-0_8.pdf) from [Real Quaternionic
Calculus Handbook](https://www.springerprofessional.de/en/real-
quaternionic-calculus-handbook/1981012) which I don't have access to
via our library for other chapters, although Google books has a
[reasonable coverage of
it](https://books.google.com/books?id=YnS8BAAAQBAJ&printsec=frontcover&
source=gbs_ge_summary_r&cad=0#v=onepage&q&f=false).

GitHub has been sick, apologies for all those (and this one which I need to add back because it was not handling deletes well either).

Reply:
I don't have access either, and I can't find a refernce to the arc functions. They look pretty reasonable though.

Reply:
I'll file an issue to improve thetesting when wecan.

Reply:
I've found a copy of the book and have added at least a test for Sin and Cos, but I suspect there are errors in the book.

If I use the literal translation of definition for sin given by them (using our functions)
```
	sgn := unit(uv)
	return Scale(-0.5, (Mul(sgn, Sub(Exp(Mul(q, sgn)), Exp(Scale(-1, (Mul(q, sgn))))))))
```
I get the same result for the tests as with the current code, but it disagrees with an exercise solution by replacement of sin for cos in one term.

They give sin(1 + i + j + k) = -sin 1 sinh sqrt 3 + (cos 1 sinh sqrt 3)/sqrt 3 (i + j + k), but this does not agree with the code above (though with the real term being -sin 1 cosh sqrt 3 it does).